examples/java8/src/test/java/org/apache/beam/examples/SnippetsTest.java

```diff
@@ -0,0 +1,111 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.examples;
+
+import java.io.IOException;
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import org.apache.beam.sdk.testing.PAssert;
+import org.apache.beam.sdk.testing.TestPipeline;
+import org.apache.beam.sdk.transforms.Create;
+import org.apache.beam.sdk.transforms.join.CoGbkResult;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TupleTag;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+
+/**
+ * Tests for Snippets.
+ */
+@RunWith(JUnit4.class)
+public class SnippetsTest implements Serializable {
+
+  @Rule
+  public transient TestPipeline p = TestPipeline.create();
+
+  /* Tests CoGroupByKeyTuple */
+  @Test
+  public void testCoGroupByKeyTuple() throws IOException {
+    // [START CoGroupByKeyTupleInputs]
+    final List<KV<String, String>> emailList = Arrays.asList(
+        KV.of("amy", "amy@example.com"),
+        KV.of("carl", "carl@example.com"),
+        KV.of("julia", "julia@example.com"),
+        KV.of("carl", "carl@email.com"));
+
+    final List<KV<String, String>> phoneList = Arrays.asList(
+        KV.of("amy", "111-222-3333"),
+        KV.of("james", "222-333-4444"),
+        KV.of("amy", "333-444-5555"),
+        KV.of("carl", "444-555-6666"));
+    // [END CoGroupByKeyTupleInputs]
+
+    // [START CoGroupByKeyTupleOutputs]
+    final TupleTag<String> emailsTag = new TupleTag();
+    final TupleTag<String> phonesTag = new TupleTag();
+
+    final List<KV<String, CoGbkResult>> expectedResults = Arrays.asList(
+        KV.of("amy", CoGbkResult
+          .of(emailsTag, Arrays.asList("amy@example.com"))
+          .and(phonesTag, Arrays.asList("111-222-3333", "333-444-5555"))),
+        KV.of("carl", CoGbkResult
+          .of(emailsTag, Arrays.asList("carl@email.com", "carl@example.com"))
+          .and(phonesTag, Arrays.asList("444-555-6666"))),
+        KV.of("james", CoGbkResult
+          .of(emailsTag, Arrays.asList())
+          .and(phonesTag, Arrays.asList("222-333-4444"))),
+        KV.of("julia", CoGbkResult
+          .of(emailsTag, Arrays.asList("julia@example.com"))
+          .and(phonesTag, Arrays.asList())));
+    // [END CoGroupByKeyTupleOutputs]
+
+    PCollection<String> actualFormattedResults =
+        Snippets.coGroupByKeyTuple(p, emailsTag, phonesTag, emailList, phoneList);
+
+    // [START CoGroupByKeyTupleFormattedOutputs]
+    final List<String> formattedResults = Arrays.asList(
+        "amy; ['amy@example.com']; ['111-222-3333', '333-444-5555']",
+        "carl; ['carl@email.com', 'carl@example.com']; ['444-555-6666']",
```

Comment:
<!--new_thread; commit:b7cd18485fb9d3c89bc5177ad8ab8742faa39c06; resolved:0-->
The order of iterables produced by a CoGroupByKey is not stable, and as such (and because you don't sort the things before putting them in the list, or use an ordered data structure) this test is overspecified

Consider using KvMatcher if possible

Reply:
The results of CoGroupByKey are being sorted in Snippets.formatCoGbkResults() which is the helper function to transform the results into these strings.